docs/unmanaged_configuration.md

```diff
@@ -25,3 +25,50 @@ oc set env dc/fluentd MERGE_JSON_LOG=true
 * Possible log loss due to Elasticsearch rejecting documents due to inconsistent type mappings
 * Potential buffer storage leak caused by rejected message cycling
 * Overwrite of data for field with same names
+
+## ElasticSearch
+### openshift_logging_es_allow_external
+If openshift_logging_es_allow_external is set to True, Elasticsearch was exposed as a Route.
+The following steps will achieve the goal.
+* Create a route for ElasticSearch.
+```
+$ oc create -f - << EOF
+apiVersion: route.openshift.io/v1
+kind: Route
+metadata:
+  name: elasticsearch
+  namespace: openshift-logging
+spec:
+  host: elasticsearch.openshift-logging.svc.cluster.local
```

Comment:
If we leave this blank, the platform will assign it a fully qualified name using its rules.  This means there will be no need for the steps to get the canonical name, or populate /etc/hosts

Reply:
@jcantrill, @richm, Could you please help me?
I restarted with this yaml file.
http://pastebin.test.redhat.com/713292
It seems it's badly broken... :(
If I leave the host blank, then the spec.host is filled with this FQDN (note: ntest0 is a cluster name.)
```
  spec:
    host: elasticsearch-openshift-logging.apps.ntest0.devcluster.openshift.com
```
Then, my attempt to search in ElasticSearch with curl failed with this error.
curl: (51) SSL: certificate subject name 'system.admin' does not match target host name 'elasticsearch-openshift-logging.apps.ntest0.devcluster.openshift.com'

I'd guess my spec.tls.key, spec.tls.certificate, spec.tls.caCertificate, and spec.tls.destinationCACertificate should be wrong.  I put admin-key, admin-cert, admin-ca and admin-ca from elasticsearch secret there, individually. 

Looking at the ansible role, es_key is from system.logging.es.key,  es_cert is from system.logging.es.crt, and es_ca is from ca.crt.  Are they different from admin-key, -cert and -ca by any chance?

Also, according to the following task, it seems we need to generate destinationCACertificate.  But it's not clear to me how to do it manually...
https://github.com/openshift/openshift-ansible/blob/master/roles/openshift_logging_elasticsearch/tasks/main.yaml#L470-L478

Thanks, in advance...

Reply:
> @jcantrill, @richm, Could you please help me?
> I restarted with this yaml file.
> http://pastebin.test.redhat.com/713292
> It seems it's badly broken... :(
> If I leave the host blank, then the spec.host is filled with this FQDN (note: ntest0 is a cluster name.)
> 
> ```
>   spec:
>     host: elasticsearch-openshift-logging.apps.ntest0.devcluster.openshift.com
> ```

Yes, you must have a host name.  In the ansible code, you must provide the hostname in openshift_logging_es_hostname (or use the default value): https://github.com/openshift/openshift-ansible/blob/release-3.11/roles/openshift_logging/defaults/main.yml#L118

This hostname is used by the router to figure out which service to use.  That is - all of the connections are coming into the router host + port.  The router looks at the requested hostname to figure out which service to send the request to.

I guess the route code auto-generates a hostname if you do not provide one.

> 
> Then, my attempt to search in ElasticSearch with curl failed with this error.
> curl: (51) SSL: certificate subject name 'system.admin' does not match target host name 'elasticsearch-openshift-logging.apps.ntest0.devcluster.openshift.com'
> 
> I'd guess my spec.tls.key, spec.tls.certificate, spec.tls.caCertificate, and spec.tls.destinationCACertificate should be wrong. I put admin-key, admin-cert, admin-ca and admin-ca from elasticsearch secret there, individually.

tls.key and tls.certificate must be the _server cert_, not admin-cert which is a _client_ cert.

> 
> Looking at the ansible role, es_key is from system.logging.es.key, es_cert is from system.logging.es.crt, and es_ca is from ca.crt. Are they different from admin-key, -cert and -ca by any chance?

Yes, they are server certs.

> 
> Also, according to the following task, it seems we need to generate destinationCACertificate. But it's not clear to me how to do it manually...
> https://github.com/openshift/openshift-ansible/blob/master/roles/openshift_logging_elasticsearch/tasks/main.yaml#L470-L478
> 
> Thanks, in advance...

The destinationCACertificate is the CA cert for the CA that issued the elasticsearch server certs.

The way re-encrypt routes work is this:

client -> router -> elasticsearch

The router terminates the TLS connection, opens a new connection to elasticsearch, and forwards the request to elasticsearch.  Also note that you cannot use client cert auth with a router re-encrypt route - for client cert auth to work, the TLS termination must be done by Elasticsearch, which can be done with a passthrough route, or by using externalIPs on the service.  The original reason we added support for openshift_logging_es_allow_external was for CloudForms to use token based auth to access Elasticsearch from outside of the cluster, which is accomplished by a re-encrypt route.

Elasticsearch has a TLS server cert and key, issued by the logging CA.  The router also has a TLS server cert and key, issued by the logging CA (in the openshift-ansible code).  I think the router TLS server cert and key can be the same as the Elasticsearch TLS server cert key, not sure how the openshift-ansible code works.

Reply:
I have to look at the arguments for the command again to see if it can cover all gaps, but have we thought about providing an `oc expose` command for the elasticsearch service instead of a template file?

Reply:
Thank you, @richm, @ewolinetz.

I think I extracted the server key and cert and put them in the spec.tls.key and cert.  But my attempt to search against the exposed elasticsearch fails with "Authentication finally failed"...  Does this mean the key and cert to "reencrypt" (router -> elasticsearch) are broken?  (Or is this curl command line not appropriate to test the exposed elasticsearch addr?)
```
$ curl --cacert ./admin-ca --cert ./logging-es.crt --key ./logging-es.key https://${ESFQDN}:443/.operations.*/_search
Authentication finally failed
$ curl --cacert ./admin-ca --cert ./admin-cert --key ./admin-key https://${ESFQDN}:443/.operations.*/_search
Authentication finally failed
```

Reply:
> Thank you, @richm, @ewolinetz.
> 
> I think I extracted the server key and cert and put them in the spec.tls.key and cert. But my attempt to search against the exposed elasticsearch fails with "Authentication finally failed"... Does this mean the key and cert to "reencrypt" (router -> elasticsearch) are broken? (Or is this curl command line not appropriate to test the exposed elasticsearch addr?)
> 
> ```
> $ curl --cacert ./admin-ca --cert ./logging-es.crt --key ./logging-es.key https://${ESFQDN}:443/.operations.*/_search
> Authentication finally failed
> $ curl --cacert ./admin-ca --cert ./admin-cert --key ./admin-key https://${ESFQDN}:443/.operations.*/_search
> Authentication finally failed
> ```

You cannot use client cert/key auth with a re-encrypt route.  You'll have to use token based auth e.g. see https://github.com/openshift/origin-aggregated-logging/blob/master/hack/testing/util.sh#L210

Reply:
Thanks a lot, @richm !!  The following command line [1] returns the search results successfully!  One question remains...  The curl_es_with_token function hardcodes the port 9200 for the elasticsearch.  But my test hangs if I put 9200, while 443 works.  Is it something I should put in the elasticsearch route manifest?  Or is it ok if it works with 443?

[1]
`curl --silent --insecure -H "Authorization: Bearer $( oc whoami -t )" "https://elasticsearch.openshift-logging.svc.cluster.local:443/.operations.*/_search" | jq`

Reply:
Another note: as @jcantrill suggested, without specifying spec.host in the original manifest nor adding IP FQDN to the local /etc/hosts, the command line to verify works!  It makes the steps simpler. :)
`curl --silent --insecure -H "Authorization: Bearer $( oc whoami -t )" "https://$( oc get route elasticsearch -o jsonpath='{.spec.host}' ):443/.operations.*/_search" | jq`

Reply:
When using a route, you have to use port `443` (or just leave it off - when using a url which begins with `https`, the port `443` is used by default).  The router listens on port `443`, and redirects your request based on the hostname to `eshost:9200`